### PR TITLE
[task-assign] Assign GitHub users to tasks

### PR DESCRIPTION
**Pull request by wlonk**
cat

### CODE REVIEW

**Comment by jgerigmeyer** [comment]
[I want to assign team member(s) to a specific task.](https://trello.com/c/TbBcu3Zu/22-i-want-to-assign-team-members-to-a-specific-task)

**Comment by jgerigmeyer** [comment]
@wlonk I think we want to add a similar validator here, such that `github_users` can only be added to a Task if already in the parent Project.

**Comment by wlonk** [comment]
Yep, that's on my agenda and why I made sure to do it as a pluggable validator from the get-go.

File: src/js/components/commits/list.tsx [hunk 1]
```diff
@@ -39,8 +39,9 @@ const AuthorTableCell = ({ item, className, ...props }: TableCellProps) => {
     return null;
   }
   let author = item.author.username;
-  if (item.author.name && item.author.name !== author) {
-    author = `${author} (${item.author.name})`;
+  const name = item.author?.name;
```
Comment (duggiemitchell) [inline src/js/components/commits/list.tsx]: Ooh fancy.

File: metashare/api/validators.py [hunk 2]
```diff
@@ -49,9 +49,13 @@ def filter_queryset(self, attrs, queryset, serializer):
...
+class GitHubUserValidator:
+    def __init__(self, *, parent):
+        self.parent = parent
+
+    def __call__(self, cleaned_data):
+        parent = cleaned_data.get(self.parent)
+        github_users = cleaned_data.get("github_users", [])
+        for user in github_users:
+            if user not in parent.github_users:
+                raise ValidationError(f"Invalid github_users value: {user}")
```
Comment (jgerigmeyer) [inline metashare/api/validators.py]: @wlonk This validator is still (again) only used for projects, so we could revert this change if you like.

**Comment by duggiemitchell** [comment]
I like a lot of the changes in the Org Cards.  Should these changes be merged into the PR for `not-latest-commit` since it's dealing with the same component?

File: metashare/api/serializers.py [hunk 3]
```diff
@@ -188,6 +181,8 @@ class Meta:
             "pr_url",
             "status",
             "pr_is_open",
+            "assigned_dev",
+            "assigned_qa",
```
Comment (jgerigmeyer) [inline metashare/api/serializers.py]: @wlonk I think we maybe need a validator for these, making sure the user is already in `project.github_users`? Tho that might depend on the answer to my first question here: https://trello.com/c/TbBcu3Zu/22-i-want-to-assign-team-members-to-a-specific-task#comment-5e2e156a2623a30e7fdad423

That is, if a user is assigned to a Task, _then_ removed from the Project -- I'm not sure if we want subsequent updates to the Task to fail validation or not...

File: src/js/components/commits/list.tsx (hunk 1 above)
Comment (jgerigmeyer) [reply src/js/components/commits/list.tsx]: It's really all your fault... https://oddbird.slack.com/archives/C1310AQNA/p1577730958000500 😄

**Comment by jgerigmeyer** [comment]
> I like a lot of the changes in the Org Cards. Should these changes be merged into the PR for `not-latest-commit` since it's dealing with the same component?

Yes, I think eventually this branch should be merged into `not-latest-commit` and `task-pr-status`. I'm happy to do that merge if you'd like?

File: metashare/api/validators.py (hunk 2 above)
Comment (wlonk) [reply metashare/api/validators.py]: I'm fine with this as-is. The generalization was motivated by code, even if the code is pulled back from there now, and so we understand the problem space enough to justify it.

File: metashare/api/serializers.py (hunk 3 above)
Comment (wlonk) [reply metashare/api/serializers.py]: Right. I think we _don't_ but I wanna talk this over with @davisagli at least.

Comment (jgerigmeyer) [reply metashare/api/serializers.py]: Ok... in that case we could validate only when a user is added, or not at all on the back end, and allow the front end to limit the available options.

**Comment by jgerigmeyer** [comment]
@wlonk As I think about it more, I think it would make sense to have the back end enforce "0 or 1 dev orgs per task" and "0 or 1 qa orgs per task". It seems like a simple thing to enforce on org creation, and could avoid weird potential bug edge cases.

**Comment by wlonk** [comment]
@jgerigmeyer sounds good to me.

File: metashare/api/validators.py [hunk 4]
```diff
@@ -49,9 +49,13 @@ def filter_queryset(self, attrs, queryset, serializer):
...
-    for user in github_users:
-        if user not in repository.github_users:
-            raise ValidationError(f"Invalid github_users value: {user}")
+class GitHubUserValidator:
+    def __init__(self, *, parent):
+        self.parent = parent
+
+    def __call__(self, cleaned_data):
+        parent = cleaned_data.get(self.parent)
+        github_users = cleaned_data.get("github_users", [])
```
[the remark this answers is not in the document]
Comment (wlonk) [reply metashare/api/validators.py]: I don't think we allow for partial-updates (whether `PUT` or the arguably-more-proper `PATCH`) per our public API anyway, so if I'm right about that, I'm not concerned about unexpected behavior if someone tries it.

File: metashare/api/validators.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (wlonk) [reply metashare/api/validators.py]: Yes.

File: metashare/api/serializers.py [hunk 5]
```diff
@@ -287,6 +283,15 @@ class Meta:
     def get_has_unsaved_changes(self, obj) -> bool:
         return bool(getattr(obj, "unsaved_changes", {}))
 
+    def validate(self, data):
+        if ScratchOrg.objects.filter(
+            task=data["task"], org_type=data["org_type"]
+        ).exists():
+            raise serializers.ValidationError(
+                "A ScratchOrg of this type already exists for this task."
```
[the remark this answers is not in the document]
Comment (wlonk) [reply metashare/api/serializers.py]: Yes.

File: metashare/api/migrations/0048_scratchorg_owner_username.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name="scratchorg",
+            name="owner_username",
+            field=sfdo_template_helpers.fields.string.StringField(blank=True),
+        ),
+        migrations.RunPython(set_org_owner_username, migrations.RunPython.noop),
+    ]
```
[the remark this answers is not in the document]
Comment (wlonk) [reply metashare/api/migrations/0048_scratchorg_owner_username.py]: Noted!

File: metashare/api/validators.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jgerigmeyer) [reply metashare/api/validators.py]: Okay. We do allow for partial `POST`, but we require the `parent` in that case. I'm okay with it if you are.

File: metashare/api/views.py [hunk 7]
```diff
@@ -221,13 +217,14 @@ def retrieve(self, request, *args, **kwargs):
...
         conditions = [
-            instance.owner == request.user,
             instance.org_type == SCRATCH_ORG_TYPES.Dev,
             instance.url is not None,
             instance.delete_queued_at is None,
             not instance.currently_capturing_changes,
             not instance.currently_refreshing_changes,
         ]
+        if not request.GET.get("get_unsaved_changes"):
+            conditions.append(instance.owner == request.user)
```
Comment (jgerigmeyer) [inline metashare/api/views.py]: @wlonk I did not add tests for these changes because coverage was still at 100%, but feel free to if you like.

File: metashare/api/views.py [hunk 8]
```diff
@@ -198,13 +193,14 @@ def list(self, request, *args, **kwargs):
...
         # worker queues easily this way:
         filters = {
-            "owner": request.user,
             "org_type": SCRATCH_ORG_TYPES.Dev,
             "url__isnull": False,
             "delete_queued_at__isnull": True,
             "currently_capturing_changes": False,
             "currently_refreshing_changes": False,
         }
+        if not request.GET.get("get_unsaved_changes"):
```
Comment (wlonk) [inline metashare/api/views.py]:
```suggestion
        if not request.query_params.get("get_unsaved_changes"):
```
See https://www.django-rest-framework.org/api-guide/requests/#query_params

File: metashare/api/views.py (hunk 7 above)
Comment (wlonk) [reply metashare/api/views.py]: I'm OK with not changing the tests; we don't have branch coverage set up in Python at this point. We maybe should, but that's another issue.

Also, make the change to `query_params` here too.

File: metashare/api/models.py [hunk 9]
```diff
@@ -502,6 +504,7 @@ class ScratchOrg(PushMixin, HashIdMixin, TimestampsMixin, models.Model):
     config = JSONField(default=dict, encoder=DjangoJSONEncoder, blank=True)
     delete_queued_at = models.DateTimeField(null=True, blank=True)
     owner_sf_id = StringField(blank=True)
+    owner_username = StringField(blank=True)
```
Comment (davisagli) [inline metashare/api/models.py]: I wish this was clearer that this is a _github_ username, and that owner_sf_id is actually a Salesforce username rather than an Id.

File: metashare/api/serializers.py [hunk 10]
```diff
@@ -287,6 +283,15 @@ class Meta:
...
         return bool(getattr(obj, "unsaved_changes", {}))
 
+    def validate(self, data):
+        if ScratchOrg.objects.filter(
+            task=data["task"], org_type=data["org_type"]
+        ).exists():
+            raise serializers.ValidationError(
+                _("A ScratchOrg of this type already exists for this task.")
+            )
+        return data
```
Comment (davisagli) [inline metashare/api/serializers.py]: Just keep in mind, I would guess that at _some_ point in the future we'll want to allow for multiple reviewers on one task, each with their own org.

File: src/js/components/tasks/cards/orgActions.tsx [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+      <Dropdown
+        align="right"
+        assistiveText={{ icon: i18n.t('Org Actions') }}
+        buttonClassName="slds-button_icon-x-small"
+        buttonVariant="icon"
+        iconCategory="utility"
+        iconName="down"
+        iconSize="small"
+        iconVariant="border-filled"
+        width="xx-small"
```
Comment (stacyk) [inline src/js/components/tasks/cards/orgActions.tsx]: I'm not sure we need the xx-small width since that is the default min-width of a dropdown. I don't know why it is even an option.

File: src/js/components/tasks/cards/orgActions.tsx [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        options={[{ id: 0, label: i18n.t('Delete Org') }]}
```
Comment (stacyk) [inline src/js/components/tasks/cards/orgActions.tsx]: maybe this is a react thing we can't avoid, and I'm not sure if this comment is in the right place, but "Delete Org" is the link text and the link title attribute which is redundant for screen readers if the title is exactly the same or doesn't provide additional useful info.

File: src/js/components/tasks/cards/card.tsx [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+          <div className="slds-m-bottom_small">
+            <UserCard user={assignedUser} className="nested-card" />
+          </div>
+        )}
+        {(assignedUser || ownedByWrongUser) && (
+          <>
+            <hr className="slds-m-vertical_none" />
+            <Card
+              className="nested-card"
+              bodyClassName="slds-card__body_inner"
```
Comment (stacyk) [inline src/js/components/tasks/cards/card.tsx]: I don't think we want this here as it adds an indent to the info. I'm not sure if it only adds padding, or if it does additional things that are necessary. If it is padding only, we can remove it.

File: src/js/components/tasks/cards/confirmRemoveUserModal.tsx [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+    handleClose();
+    /* istanbul ignore else */
+    if (org && assignedUser) {
+      handleDelete(org, { ...assignedUser });
+    }
+  };
+  const heading =
+    type === ORG_TYPES.QA
+      ? i18n.t('Confirm Change Reviewer and Delete Review Org')
+      : i18n.t('Confirm Change Developer and Delete Dev Org');
```
Comment (stacyk) [inline src/js/components/tasks/cards/confirmRemoveUserModal.tsx]: Could these titles say "Confirm Reviewer Change ..." or "Confirm Changing the Reviewer ..." it reads a little funny as its currently written IMO. Same for Developer

File: src/js/components/tasks/cards/orgActions.tsx (hunk 11 above)
Comment (stacyk) [reply src/js/components/tasks/cards/orgActions.tsx]: nope, definitely wrong about that one., ha ha

File: src/js/components/user/githubUser.tsx [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+            </div>
+            <GitHubUserButton user={selectedUser} isSelected />
+          </div>
+          <hr className="slds-m-vertical_none slds-m-horizontal_small" />
+        </>
+      )}
+      {filteredUsers.length ? (
+        <div className="slds-p-around_small">
+          <div className="slds-text-title slds-m-bottom_xx-small">
+            {i18n.t('Assign To User')}
```
Comment (stacyk) [inline src/js/components/user/githubUser.tsx]: Should we use the label collaborator here? Assign to  `developer` (or reviewer`) ... or generically "Change assignment to:"

File: src/js/components/user/githubUser.tsx [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+);
+
+const UserTableCell = ({ item, ...props }: TableCellProps) => {
+  /* istanbul ignore if */
+  if (!item) {
+    return null;
+  }
+  const { login } = item;
+  return (
+    <DataTableCell {...props} title={login} className="collaborator-grid">
```
Comment (stacyk) [inline src/js/components/user/githubUser.tsx]: Is it possible to use this avatar and username as a target to toggle the checkbox?

File: src/js/components/tasks/cards/orgActions.tsx (hunk 12 above)
Comment (jgerigmeyer) [reply src/js/components/tasks/cards/orgActions.tsx]: Yeah, that's SLDS and matches the markup in SLDS, not something we control: https://lightningdesignsystem.com/components/menus/#Base

File: src/js/components/user/githubUser.tsx (hunk 15 above)
Comment (jgerigmeyer) [reply src/js/components/user/githubUser.tsx]: I changed to "Assign to GitHub User" so that it's clear these are GitHub usernames. Does that seem better?

File: src/js/components/user/githubUser.tsx (hunk 16 above)
Comment (jgerigmeyer) [reply src/js/components/user/githubUser.tsx]: @stacyk Yes -- as part of that I brought this styling in line with the other modal a bit more. Feel free to review and cleanup, or let me know if there are minor changes you'd like (e.g. remove the outline on hover?).

Screenshot 2020-01-30 17 46 40

**Comment by jgerigmeyer** [comment]
@stacyk I think I addressed your review comments -- please take a look.

> I was able to create an org on a task, then go to the project page and remove myself from the project after my org was created without a modal. When I tried to remove myself from a task, it gave me a confirm modal. I think I should see that same modal at the project level?

That's intentional, at least for now. We're allowing removing users from projects, even if they are still assigned to tasks within that project.

File: src/js/components/user/githubUser.tsx (hunk 16 above)
Comment (stacyk) [reply src/js/components/user/githubUser.tsx]: @jgerigmeyer I removed outline on hover, removed focus on button, added table row hover (it feels more like a hover state, but I know the people working on this will know it is just a table style). I'd like to add cursor to pointer on checkbox but I know we decided not to go down that path. I don't like the double tabbing, but I know I asked for the name to be a clickable label, and because this is in a table we can't wrap the label around the checkbox and text. What do you think of the changes I just made?

File: src/sass/patterns/_data-tables.scss [hunk 17]
```diff
@@ -19,3 +19,18 @@
...
+
+  &.slds-table_cell-buffer tr > th:first-child,
+  &.slds-table_cell-buffer tr > td:first-child,
+  &.slds-table--cell-buffer tr > th:first-child,
+  &.slds-table--cell-buffer tr > td:first-child,
+  &.slds-table th:first-child,
+  &.slds-table td:first-child {
+    padding: 0;
+  }
+}
```
Comment (jgerigmeyer) [inline src/sass/patterns/_data-tables.scss]: @stacyk I'm not sure this entire block is worth it -- looks like it just shifts the padding slightly? Or if you do think it's important, can you add a comment explaining what it does and why?

Default styles:

Screenshot 2020-01-31 15 19 41

With these overrides:

Screenshot 2020-01-31 15 19 57

Comment (stacyk) [reply src/sass/patterns/_data-tables.scss]: Ill add a comment. It is making the target area wider. So anywhere you click within a row now selects the checkbox. There used to be big gaps on the left where you couldn't click and the button on the right filled most, but not all of the space.

**Comment by jgerigmeyer** [comment]
@wlonk @duggiemitchell Have you two had a chance to review the code changes in this PR yet? Would you prefer to walk through it together?

**Comment by wlonk** [comment]
@jgerigmeyer that backend code looks fine. 👍

File: src/sass/components/_collaborators.scss [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+    }
+  }
+
+  // don't show a border on tables with checkboxes
+  .table-row-targets & {
+    &:hover {
+      border-color: transparent;
+    }
+
+    &:focus {
```
Comment (mirisuzanne) [inline src/sass/components/_collaborators.scss]: These are currently keyboard-focus targets in the tab-order,  with no focus style

Comment (jgerigmeyer) [reply src/sass/components/_collaborators.scss]: Yeah, I think @stacyk was aware of that and did it intentionally?

> I don't like the double tabbing, but I know I asked for the name to be a clickable label, and because this is in a table we can't wrap the label around the checkbox and text.

Comment (mirisuzanne) [reply src/sass/components/_collaborators.scss]: Seems like if we're going to leave the double-tab there should be *some* focus style that goes with it?

File: src/sass/patterns/_data-tables.scss [hunk 19]
```diff
@@ -19,3 +19,21 @@
...
+  }
+
+  .slds-checkbox__label {
+    padding: $spacing-medium $spacing-large;
+  }
+
+  &.slds-table_cell-buffer tr > th:first-child,
+  &.slds-table_cell-buffer tr > td:first-child,
+  &.slds-table th:first-child,
+  &.slds-table td:first-child {
```
Comment (jgerigmeyer) [inline src/sass/patterns/_data-tables.scss]: As far as I can tell `.slds-table_cell-buffer` is only applied to `.slds-table`, so I don't see how these first two definitions could differ from the last two. Can we just remove the `.slds-table_cell-buffer` ones?

File: src/sass/components/_collaborators.scss (hunk 18 above)
Comment (jgerigmeyer) [reply src/sass/components/_collaborators.scss]: Added focus styles back.

Screenshot 2020-01-31 21 59 02

**Comment by jgerigmeyer** [comment]
@stacyk @mirisuzanne Merging, but happy for post-merge review/cleanup as needed.
